src/ir_Funiki.cpp

```diff
@@ -0,0 +1,580 @@
+    // Block #1
+    sendGeneric(kFunikiHdrMark, kFunikiHdrSpace, kFunikiBitMark,
+                kFunikiOneSpace,
+                kFunikiBitMark, kFunikiZeroSpace, 0, 0,  // No Footer.
+                data, nbytes, 38, false, 0, 50);
+    // Footer #1
+    sendGeneric(0, 0,  // No Header
+                kFunikiBitMark, kFunikiOneSpace, kFunikiBitMark,
+                kFunikiZeroSpace,
+                kFunikiBitMark, kFunikiMsgSpace, 0b101, 3, 38, false, 0, 50);
+  }
+}
+#endif  // SEND_FUNIKI
+
+/// Class constructor
+/// @param[in] pin GPIO to be used when sending.
+/// @param[in] model The enum of the model to be emulated.
+/// @param[in] inverted Is the output signal to be inverted?
+/// @param[in] use_modulation Is frequency modulation to be used?
+IRFunikiAC::IRFunikiAC(const uint16_t pin, const funiki_ac_remote_model_t model,
+                   const bool inverted, const bool use_modulation)
+    : _irsend(pin, inverted, use_modulation) {
+  stateReset();
+  setModel(model);
+}
+
+/// Reset the internal state to a fixed known good state.
+void IRFunikiAC::stateReset(void) {
+  // This resets to a known-good state to Power Off, Fan Auto, Mode Auto, 25C.
+  std::memset(_.remote_state, 0, sizeof _.remote_state);
+  _.Temp = 9;  // _.remote_state[1] = 0x09;
+  _.unknown1 = 0x60;  // _.remote_state[5] = 0x20;
+}
+
+/// Fix up the internal state so it is correct.
+/// @note Internal use only.
+void IRFunikiAC::fixup(void) {
+  setPower(getPower());  // Redo the power bits as they differ between models.
+  checksum();  // Calculate the checksums
+}
+
+/// Set up hardware to be able to send a message.
+void IRFunikiAC::begin(void) { _irsend.begin(); }
+
+#if SEND_FUNIKI
+/// Send the current internal state as an IR message.
+/// @param[in] repeat Nr. of times the message will be repeated.
+void IRFunikiAC::send(const uint16_t repeat) {
+  _irsend.sendFuniki(getRaw(), kFunikiStateLength, repeat);
+}
+#endif  // SEND_FUNIKI
+
+/// Get a PTR to the internal state/code for this protocol.
+/// @return PTR to a code for this protocol based on the current internal state.
+uint8_t* IRFunikiAC::getRaw(void) {
+  fixup();  // Ensure correct settings before sending.
+  return _.remote_state;
+}
+
+/// Set the internal state from a valid code for this protocol.
+/// @param[in] new_code A valid code for this protocol.
+void IRFunikiAC::setRaw(const uint8_t new_code[]) {
+  std::memcpy(_.remote_state, new_code, kFunikiStateLength);
+  // We can only detect the difference between models when the power is on.
+  if (_.Power) {
+      _model = funiki_ac_remote_model_t::UNKOWN;
+  }
+}
+
+/// Calculate and set the checksum values for the internal state.
+/// @param[in] length The size/length of the state array to fix the checksum of.
+void IRFunikiAC::checksum(const uint16_t length) {
+  (void)(length);
+  // Funiki uses the same checksum alg. as Kelvinator's block checksum.
+  // _.Sum = IRKelvinatorAC::calcBlockChecksum(_.remote_state, length);
+}
```

Comment:
This doesn't make sense. If there is no checksum, then you shouldn't need this procedure.
And the `validChecksum()` function shouldn't be needed. If it is needed, the the checksum should do something.

If they are both un-needed, they should be removed.